File: my-folder/0410-split-array-largest-sum/solution.cpp
class Solution {
public:
    bool check(vector<int> nums,int maxSum,int parts){
        int part = 1;
        int current = 0;
        for(int i:nums){
            if(current+i<=maxSum){
                current+=i;
            }
            else {
                current = i;
                part++;
                }

            if(part>parts || current>maxSum) return false;
        }
        return true;
    }
    int splitArray(vector<int>& nums, int k) {
        int low =0;
        int high = accumulate(nums.begin(),nums.end(),0);
        
        while(low<=high){
            int mid = low + (high-low)/2;
            if(check(nums,mid,k)){
                high = mid-1;
            }
            else low = mid+1;
            
        } 
        return low;
        
    }
};

File: my-folder/0410-split-array-largest-sum/solution.java
class Solution {
    boolean solve(int[] nums,int k,int maxSum){
        int current = 0;
        for(int i = 0;i<nums.length;i++){
            if(current+nums[i]<=maxSum){
                current+=nums[i];
            }
            else{
                k--;
                current = nums[i];
                if(k==0) return false;
            }

        }
        return true;
    }
    public int splitArray(int[] nums, int k) {
       int left = Arrays.stream(nums).max().getAsInt();
       int right = Arrays.stream(nums).sum();
       int ans = -1;
       while(left<=right){
        int mid = (left+right)/2;
        if(solve(nums,k,mid)){
            ans = mid;
            right = mid-1;
        }
        else left = mid+1;
       }
       return ans;
        
    }
}
